### ISSUE TITLE
How do we save the finetuned ASR model and use them for inference?

### ISSUE THREAD

**Issue by Trikaldarshi** [speechbrain/speechbrain#1222]
Hello, 
I was able to fine-tune a model on my dataset as given in the tutorial [Pretrained Models and Fine-Tuning](https://colab.research.google.com/drive/1LN7R3U3xneDgDRK2gC5MzGkLysCWxuC3?usp=sharing).
However, I couldn't find a clear way of saving the fine-tuned model ( something equivalent to torch.save or pickle). I have gone through the [checkpoints](https://colab.research.google.com/drive/1VH7U0oP3CZsUNtChJT2ewbV_q1QX8xre?usp=sharing) tutorial and tried to do something as follows (steps after brain.fit() -- to save and load the fine-tune model):
```
from speechbrain.utils.checkpoints import Checkpointer
checkpoint_dir = "./full_example_checkpoints"
checkpointer = Checkpointer(checkpoint_dir)

modules = {"enc": asr_model.mods.encoder.model, 
           "emb": asr_model.hparams.emb,
           "dec": asr_model.hparams.dec,
           "compute_features": asr_model.mods.encoder.compute_features,
           "normalize": asr_model.mods.encoder.normalize,
           "seq_lin": asr_model.hparams.seq_lin, 
}
checkpointer.add_recoverables(modules)
ckpt = checkpointer.save_checkpoint()
ckpt_finder = Checkpointer(checkpoint_dir)
get_ckpt = ckpt_finder.find_checkpoint()

**% load a fresh model and then transfer the parameters of fine-tuned models into that**

new_asr_model = EncoderDecoderASR.from_hparams(source="speechbrain/asr-crdnn-rnnlm-librispeech", savedir="./pretrained_ASR")

**% parameter transfer**
sb.utils.checkpoints.torch_parameter_transfer(new_asr_model.mods.encoder.model, get_ckpt.paramfiles['enc'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(new_asr_model.hparams.emb, get_ckpt.paramfiles['emb'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(new_asr_model.hparams.dec, get_ckpt.paramfiles['dec'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(new_asr_model.mods.encoder.compute_features, best_ckpt.paramfiles['compute_features'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(new_asr_model.mods.encoder.normalize, get_ckpt.paramfiles['normalize'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(new_asr_model.hparams.seq_lin, get_ckpt.paramfiles['seq_lin'], device='cpu')
```

I believe there must be a better way to do this. Could you please confirm that above code will work? Also, it would be great if you could provide some hints related to this. Thanks in advance.

**Comment by Gastron** [comment]
I think you've mostly got it. I'll try to outline the whole process since I was not sure what part you were having trouble with.

1. During pretraining, you have one `Checkpointer` instance (used in the normal way) which saves checkpoints to be used for recovery and as the intermediate pretraining result. Let's say it saves checkpoints in `./exp/pretraining/checkpoints`. This could happen in a separate script (`pretrain.py` or something).

2. Now you would probably move to a new script (`finetune.py` for example). Here, to find the best checkpoint to load for finetuning, you can use a `Checkpointer` without actually saving/loading anything:
```python
ckpt_finder = Checkpointer("./exp/pretraining/checkpoints")
ckpt = ckpt_finder.find_checkpoint(min_key="loss")  # Or whatever you want as the criterion.
```

3. Still in the same script, to transfer the parameters you can use `sb.utils.checkpoints.torch_parameter_transfer`:
```python
sb.utils.checkpoints.torch_parameter_transfer(hparams["encoder"], ckpt.paramfiles['encoder'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(hparams["decoder"], ckpt.paramfiles['decoder'], device='cpu')
sb.utils.checkpoints.torch_parameter_transfer(hparams["normalizer"], ckpt.paramfiles['normalizer'], device='cpu')
```
Or it may be neater to use a `Pretrainer`, and create it in the YAML:
```yaml
pretrainer:
    collect_in: !ref <output_dir>/pretraining_files
    loadables:
        encoder: !ref <encoder>
        decoder: !ref <decoder>
        normalizer: !ref <normalizer>
```
```python
hparams["pretrainer"].collect_files(default_source=ckpt.path)
hparams["pretainer"].load_collected(device="cpu")
```
And if your pretrained model comes from a different source, e.g. online, you will of course skip steps 1 and 2. You could load the files here in step 3 almost the same. If you have a HuggingFace repo with the files `encoder.ckpt`, `decoder.ckpt` and `normalizer.ckpt`, say under "researcher5124/great-asr-base":
```yaml
pretrainer:
    collect_in: !ref <output_dir>/pretraining_files
    loadables:
        encoder: !ref <encoder>
        decoder: !ref <decoder>
        normalizer: !ref <normalizer>
```
```python
hparams["pretrainer"].collect_files(default_source="researcher5124/great-asr-base")
hparams["pretainer"].load_collected(device="cpu")
```

4. Now still in the finetuning script, use a `Checkpointer` (separate from the one in step 3) in the normal way (declare it in YAML, etc.) during the finetuning to save checkpoints both for recovery and for the final finetuned result.

5. Now, inference with the finetuned model. If you are just running an experiment, you should probably just do inference via the `Brain.evaluate()` in the same script as your finetuning - basically the normal way. 

6. If you want to publish the model on HuggingFace or otherwise use it via a [Pretrained interface](https://speechbrain.readthedocs.io/en/latest/API/speechbrain.pretrained.interfaces.html#speechbrain.pretrained.interfaces.Pretrained), you should essentially manually package all the files that are needed: find the right checkpoint directory, copy the model checkpoints and write a suitable hyperparams.yaml. In the hyperparams.yaml file, use a `Pretrainer` as above in step 3 - the `Pretrained.from_hparams` calls `collect_files` and `load_collected`.

**Comment by Trikaldarshi** [comment]
Thanks you very much for your quick and detailed comments. However, I am still confused with the explanation. I don't get the point as how can we simply store the fine-tuned model and then reload it for evaluation on new data. This seems to be achieved by speechbrain using two concepts: pretrainer and checkpoints. Here are the steps I did and I got stuck again on the same question "how do I save a fine-tuned  model and reload it from a fresh settings (say notebook) and then evaluate it?" I only understood the partial answer as we can save fine-tuned model in checkpoints. But, I don't get how do I use those checkpoints.
Here is what I did:

From here, https://github.com/speechbrain/speechbrain/tree/develop/recipes/LibriSpeech/ASR/transformer, I trained the model on my dataset. The script train.py which essentially fine-tunes the transformer model at https://huggingface.co/speechbrain/asr-transformer-transformerlm-librispeech on my data. The same scripts train.py saved my checkpoints as well (my transformer.yaml is exactly same except it takes my data in train.csv, test.csv and dev.csv). In the transformer.yaml, pretrainer and checkpoints everything is defined (also written here). I got improved performance  as well. I have my checkpoints as well.
But, I don't know how to exactly use that checkpoint in fresh settings (not in the same fine-tuning code)? I just somehow want to pick those checkpoints and evaluate it. 
```
checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
    checkpoints_dir: !ref <save_folder>
    recoverables:
        model: !ref <model>
        noam_scheduler: !ref <noam_annealing>
        normalizer: !ref <normalize>
        counter: !ref <epoch_counter>
        lm: !ref <lm_model>

pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
    collect_in: !ref <save_folder>
    loadables:
        lm: !ref <lm_model>
        tokenizer: !ref <tokenizer>
        model:  !ref <model>
        normalize: !ref <normalize>
    paths:
        lm: !ref <pretrained_lm_tokenizer_path>/lm.ckpt
        tokenizer: !ref <pretrained_lm_tokenizer_path>/tokenizer.ckpt
        model: !ref <pretrained_lm_tokenizer_path>/asr.ckpt
        normalize: !ref <pretrained_lm_tokenizer_path>/normalizer.ckpt
```
I have got checkpoints (you can see in this image). But I don't know how do I evaluate it? 
![image](https://user-images.githubusercontent.com/47347137/148875316-975cc8db-0ac6-4068-aff9-9b58a90712f4.png)

Could you please give some pointers?

**Comment by Trikaldarshi** [comment]
I tried the following command, but don't know whether it is correct way to do this.
asr_model = EncoderDecoderASR.from_hparams(source="./my_checkpoints",hparams_file='transformer.yaml', savedir="my_model")
here **my_checkpoints** contains the files from the checkpointer. In transformer.yaml, I have defined the checkpointer and pretrainder in following way:

pretrained_lm_tokenizer_path: speechbrain/asr-transformer-transformerlm-librispeech

checkpointer: !new:speechbrain.utils.checkpoints.Checkpointer
    checkpoints_dir: !ref <save_folder>
    recoverables:
        model: !ref <asr_model>
        normalizer: !ref <normalizer>
        counter: !ref <epoch_counter>
        lm: !ref <lm_model>

pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
   loadables:
      normalizer: !ref <normalizer>
      model: !ref <asr_model>
      lm: !ref <lm_model>
      tokenizer: !ref <tokenizer>
   paths:
      tokenizer: !ref <pretrained_lm_tokenizer_path>/tokenizer.ckpt

Since  the tokenizer is not the part of the checkpointer, I am downloading it from the huggingface hub.

Will it  load my the model with parameters with checkpointer fille I have given?

**Comment by Gastron** [comment]
> But I don't know how do I evaluate it?
Basically all SpeechBrain recipes use the Brain class to do the evaluation too. In evaluation you want to compute some test metric (like WER) over some test data - the Brain class is suitable for that (see `evaluate()`). You can do that in a Jupyter Notebook as well, of course, it's the same Python in the end. If you don't want to use a Brain class, then basically you gotta somehow implement the things that the Brain class did:
- running your model (just calling the feature computation, normalization, encoder, decoding, or whatever)
- looping over test data
- computing your evaluation metric
For running your model, you might be able to use a `Pretrained` subclass, if that's convenient.

> Will it load my the model with parameters with checkpointer fille I have given?
Yes. Looks right to me, except the `Checkpointer` seems to not be needed here? Note that you can basically apply the concepts from my previous reply's stages 2-3 again after finetuning to load your finetuned parameters elsewhere.

So maybe the underlying confusion here is that SpeechBrain does not have centralized training and evaluation scripts. Instead, SpeechBrain provides building blocks like the Brain class which you can use in your scripts to do what you want. The recipes are custom scripts like that, too - and generally they do both training and evaluation. But basically, you can expect to write some Python code to accomplish what you want. 

At some point, it might be helpful for you to just read the `Checkpointer`, `Pretrainer`, and `Pretrained` implementations - sounds like what you want to do is possible in pure PyTorch in a couple of lines, too, if you feel that the Checkpointer etc. abstractions are just getting in the way.

**Comment by Trikaldarshi** [comment]
Thank you very much for the help. I have figured out another way by uploading my .ckpt files from my checkpointer's directory to huggingfacehub. https://huggingface.co/Trikaldarshi/finetuned_british_asr_v1/tree/main
After that I used the following command:
`model=EncoderDecoderASR.from_hparams(source="Trikaldarshi/finetuned_british_asr_v1",hparams_file='transformer1.yaml',savedir="my_model")
`
Yes, you are right, checkpointer is not required in this case. I defined my pre-trainer as in the above **transformer1.yaml** file as follows:
```
pretrained_lm_tokenizer_path: Trikaldarshi/finetuned_british_asr_v1

pretrainer: !new:speechbrain.utils.parameter_transfer.Pretrainer
    loadables:
        lm: !ref <lm_model>
        tokenizer: !ref <tokenizer>
        model:  !ref <asr_model>
        normalizer: !ref <normalizer>
    paths:
        lm: !ref <pretrained_lm_tokenizer_path>/lm.ckpt
        tokenizer: !ref speechbrain/asr-transformer-transformerlm-librispeech/tokenizer.ckpt
        model: !ref <pretrained_lm_tokenizer_path>/model.ckpt
        normalizer: !ref <pretrained_lm_tokenizer_path>/normalizer.ckpt
```
I could load the model and it was fine. Again loading tokenizer from the original speechbrain model as checkpointers doesn't save tokenizers.
I hope this does what I wanted and I can load it anytime and just evaluate. Am I correct? If you could confirm, then I guess we are good to close the ticket. And again, thank you very much for all the help!

**Comment by Gastron** [comment]
Ok so you've decided to use the Pretrained interface. Looks ok, the paths provided to your Pretrainer seem a little weird, though I think it would work.

Here's how I'd change it, though:
1. In your HuggingFace repo, remove CKPT.yaml, brain.ckpt, counter.ckpt, dataloader-TRAIN.ckpt, noam_scheduler.ckpt, and optimizer.ckpt since those are not needed for evaluation.
2. Upload a copy of the tokenizer.ckpt (just in case this changes in the referenced repo).
3. Remove the paths argument (the whole block in your YAML) from Pretrainer - these locations are already the default behaviour, there's no need to specify them again.

**Comment by Trikaldarshi** [comment]
Hello,
Thank you very much for all the help. It is working fine for me now and is very convenient as well. Closing the ticket. Thanks again!.
